File: implementations/rust/ockam/ockam_command/src/service/echoer.rs
use ockam::Context;
use ockam_core::{Result, Routed, Worker};

pub struct Echoer;

pub const ECHOER_SERVICE_NAME: &str = "ECHOER";

#[ockam::worker]
impl Worker for Echoer {
    type Message = String;
    type Context = Context;

    async fn handle_message(&mut self, ctx: &mut Context, msg: Routed<String>) -> Result<()> {
        println!("Address: {}, Received: {}", ctx.address(), msg);

        // Echo the message body back on its return_route.
        ctx.send(msg.return_route(), msg.body()).await
    }
}

File: implementations/elixir/examples/get_started/echoer.exs
defmodule Echoer do
  use Ockam.Worker

  alias Ockam.Message
  alias Ockam.Worker

  @impl true
  def handle_message(message, %{address: address} = state) do
    IO.puts("Address: #{address}\t Received: #{inspect(message)}")
    Worker.route(Message.reply(message, address, Message.payload(message)), state)

    {:ok, state}
  end
end
